Firestore/core/src/firebase/firestore/remote/exponential_backoff.h

```diff
@@ -0,0 +1,118 @@
+   * backoff curve between attempts.
+   *
+   * The first attempt will be done immediately. After that, each retry will
+   * have a delay that is made up of a "base" delay which follows the
+   * exponential backoff curve, and a +/- <=50% "jitter" that is calculated and
+   * added to the base delay. This prevents clients from accidentally
+   * synchronizing their delays causing spikes of load to the backend.
+   *
+   * @param queue The queue to run operations on.
+   * @param timer_id The id to use when scheduling backoff operations on the
+   *     queue.
+   * @param backoff_factor The multiplier to use to determine the extended base
+   *     delay after each attempt.
+   * @param initial_delay The initial delay (used as the base delay on the first
+   *     retry attempt, that is, the second attempt). Note that jitter will
+   *     still be applied, so the actual delay could be as little as
+   *     `0.5*initial_delay`.
+   * @param max_delay The maximum base delay after which no further backoff is
+   *     performed. Note that jitter will still be applied, so the actual delay
+   *     could be as much as `1.5*max_delay`.
+   */
+  ExponentialBackoff(util::AsyncQueue* queue,
+                     util::TimerId timer_id,
+                     double backoff_factor,
+                     util::AsyncQueue::Milliseconds initial_delay,
+                     util::AsyncQueue::Milliseconds max_delay)
+      : queue_{queue},
+        timer_id_{timer_id},
+        backoff_factor_{backoff_factor},
+        initial_delay_{initial_delay},
+        max_delay_{max_delay} {
+  }
+
+  /**
+   * Resets the backoff delay.
+   *
+   * The very next `backoffAndRun` will have no delay. If it is called again
+   * (i.e. due to an error), `initial_delay` (plus jitter) will be used, and
+   * subsequent ones will increase according to the `backoff_factor`.
+   */
+  void Reset() {
+    current_base_ = RealSeconds{0};
+  }
+
+  /**
+   * Resets the backoff to the maximum delay (e.g. for use after
+   * a RESOURCE_EXHAUSTED error).
+   */
+  void ResetToMax() {
+    current_base_ = max_delay_;
+  }
+
+  /**
+   * Waits for `current_base` seconds, increases the delay and runs the
+   * specified operation. If there was a pending operation waiting to be run
+   * already, it will be canceled.
+   */
+  void BackoffAndRun(util::AsyncQueue::Operation&& operation);
+
+  /** Cancels any pending backoff operation scheduled via `BackoffAndRun`. */
+  void Cancel() {
+    delayed_operation_.Cancel();
+  }
+
+ private:
+  using RealSeconds = std::chrono::duration<double>;
+
+  RealSeconds GetDelayWithJitter();
+  RealSeconds ClampDelay(RealSeconds delay) const;
+
+  util::AsyncQueue* queue_;
+  util::TimerId timer_id_;
+  util::DelayedOperation delayed_operation_;
+
+  double backoff_factor_;
+  RealSeconds current_base_{0};
+  util::AsyncQueue::Milliseconds initial_delay_;
+  util::AsyncQueue::Milliseconds max_delay_;
+  util::SecureRandom secure_random_;
```

Comment:
A *secure* rng is probably overkill for this and could cause the system to loose entropy for operations that truly require one. Consider switching to a pseudo rng.

Reply:
On all the platforms we care about `SecureRandom` does the right thing, namely that it draws from the entropy pool only to seed and reseed a PRNG but otherwise does not draw directly from system entropy.

So while SecureRandom may be overkill, it's not harmful, and since it already exists, it's reasonable to continue using it here.

Reply:
@rsgowman It's a straightforward copy from the Objective-C++ code.

@wilhuff Thanks for the explanation!